builtin/logical/aws/client.go

```diff
@@ -23,91 +24,139 @@ import (
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
+// Return a slice of *aws.Config, based on descending configuration priority. STS endpoints are the only place this is used.
 // NOTE: The caller is required to ensure that b.clientMutex is at least read locked
-func (b *backend) getRootConfig(ctx context.Context, s logical.Storage, clientType string, logger hclog.Logger) (*aws.Config, error) {
-	credsConfig := &awsutil.CredentialsConfig{}
-	var endpoint string
-	var maxRetries int = aws.UseServiceDefaultRetries
+func (b *backend) getRootConfig(ctx context.Context, s logical.Storage, clientType string, logger hclog.Logger) ([]*aws.Config, error) {
+	// set fallback region (we can overwrite later)
+	fallbackRegion := os.Getenv("AWS_REGION")
+	if fallbackRegion == "" {
+		fallbackRegion = os.Getenv("AWS_DEFAULT_REGION")
+	}
+	if fallbackRegion == "" {
+		fallbackRegion = "us-east-1"
+	}
+
+	maxRetries := aws.UseServiceDefaultRetries
 
 	entry, err := s.Get(ctx, "config/root")
 	if err != nil {
 		return nil, err
 	}
-	if entry != nil {
-		var config rootConfig
-		if err := entry.DecodeJSON(&config); err != nil {
-			return nil, fmt.Errorf("error reading root configuration: %w", err)
+	var configs []*aws.Config
+
+	// I'm not sure this is a valid scenario, but the previous code had it as a case.
```

Comment:
I believe this is possible since none of the `aws/config/root` fields are required by the API as they can be either set to defaults or the environment. i.e., I don't believe it is required to explicitly call `aws/config/root` and in that case entry will be nil.

Reply:
I've updated the comment to reflect this